test/pools.jl

```diff
@@ -0,0 +1,99 @@
+using ConcurrentUtilities, Test
+
+@testset "Pools" begin
+    pool = Pool(Int; max=3)
+    # acquire an object from the pool
+    x1 = acquire(() -> 1, pool)
+    # no existing objects in the pool, so our function was called to create a new one
+    @test x1 == 1
+    # release back to the pool for reuse
+    release(pool, x1)
+    # acquire another object from the pool
+    x1 = acquire(() -> 2, pool)
+    # this time, the pool had an existing object, so our function was not called
+    @test x1 == 1
+    # but now there are no objects to reuse again, so the next acquire will call our function
+    x2 = acquire(() -> 2, pool)
+    @test x2 == 2
+    x3 = acquire(() -> 3, pool)
+    @test x3 == 3
+    # the pool is now at capacity, so the next acquire will block until an object is released
+    tsk = @async acquire(() -> 4, pool; forcenew=true)
+    yield()
+    @test !istaskdone(tsk)
+    # release an object back to the pool
+    release(pool, x1)
+    # now the acquire can complete
+    x1 = fetch(tsk)
+    # even though we released 1 for reuse, we passed forcenew, so our function was called to create new
+    @test x1 == 4
+    # error to try and provide a key to a non-keyed pool
+    @test_throws ArgumentError acquire(() -> 1, pool, 1)
+    # release objects back to the pool
+    release(pool, x1)
+    release(pool, x2)
+    release(pool, x3)
+    # acquire an object, but checking isvalid
+    x1 = acquire(() -> 5, pool; isvalid=x -> x == 1)
+    @test x1 == 1
+    # no valid objects, so our function was called to create a new one
+    x2 = acquire(() -> 6, pool; isvalid=x -> x == 1)
+    @test x2 == 6
+    # we have one slot left in the pool, we now throw while creating new
+    # and we want to test that the permit isn't permanently lost for the pool
+    @test_throws ErrorException acquire(() -> error("oops"), pool; forcenew=true)
+    # we can still acquire a new object
+    x3 = acquire(() -> 7, pool; forcenew=true)
+    @test x3 == 7
+    # release objects back to the pool
+    release(pool, x1)
+    release(pool, x2)
+    release(pool, x3)
+    # try to do an invalid release
+    @test_throws ErrorException release(pool, 10)
+    # test that the invalid release didn't push the object to our pool for reuse
+    x1 = acquire(() -> 8, pool)
+    @test x1 == 7
+    # calling empty! removes all objects for reuse
+    empty!(pool)
```

Comment:
why do we support `empty!` ? 

At first i thought `Pool` was supporting the exact same interface as `Semaphone`, which is just `acquire`/`release`... but then there's `empty!`

i think i'm confused by what the contract of `empty!` is supposed to be here (see my comment above). 
what does "removes all objects _for reuse_" mean? I wonder if `empty!` is even the right name.

It also makes me wonder if there should also be a `release_all(pool)` that just releases all the permits?

And if there should be other Dict-like interface functions (e.g. `keys`/`haskey` or `length`)?

Reply:
Yeah, let's come up with a better name; HTTP has this function `closeall`, but that's not quite the right thing I think for a generic pool. Maybe `evictall`? `empty!` felt the most right to me, but as you pointed out there's this tension between whether a `Pool` is just a semaphore or a Dict storing non-checked-out objects.

Reply:
Changed the function name to `drain!` in the PR